org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModelManager.java

```diff
@@ -1244,6 +1244,73 @@ public static IJavaElement determineIfOnClasspath(IResource resource, IJavaProje
 	 */
 	protected WeakHashMap<AbstractSearchScope, ?> searchScopes = new WeakHashMap<>();
 
+	/**
+	 * Current secondary types cache and temporary indexing cache
+	 *
+	 * @param secondaryTypes
+	 *            the main cache containing known secondary types in a project, can be null to indicate no search were
+	 *            done. Note, the cache might be not complete if the search is still running.
+	 * @param indexingSecondaryCache
+	 *            the temporary structure used while indexing, previously known as INDEXED_SECONDARY_TYPES entry. If it
+	 *            is null, no cache updates are running now
+	 */
+	private static record SecondaryTypesCache(Hashtable<String, Map<String, IType>> secondaryTypes,
```

Comment:
I cannot seem to find how the nested maps are guarded against concurrent reads and modifications, but I’m traveling these days and only see the code on my phone. I’ll review in depth in two weeks time.

Reply:
They don't, but the manipulations on two map *references* is now atomic (which was not the case before, and most of the logic was based on "if map 1 is null, do something with map 2" etc).